File: PAT_Basic_Level_by_cpp/1018 锤子剪刀布/1018 锤子剪刀布.cpp
// 1018 锤子剪刀布
// 统计甲乙交锋的胜、平、负次数，并输出各自获胜次数最多的手势（并列时取字母序最小者）。
// 胜负：布(B) > 锤(C) > 剪(J) > 布(B)
//
// 【实现原理】
// 算法思路：
//   1. 对于每轮出拳，判断胜负平，更新对应计数器。
//   2. 同时记录甲和乙各自以每种手势获胜的次数。
//   3. 胜负判定规则：B > C > J > B（循环克制关系）。
//   4. 最后找出各自获胜次数最多的手势：
//      - 使用 lambda 表达式遍历手势计数数组，找最大值索引。
//      - 由于数组下标 0=B, 1=C, 2=J 已按字母序排列，更新条件用严格大于，
//        当出现并列时，靠前的（字母序小的）会被保留。
// 关键点：
//   - 手势映射为下标时按字母序 B < C < J 排列，便于处理并列情况。
// 时间复杂度：O(N)，一次遍历即可。
// 空间复杂度：O(1)。
#include <iostream>

using namespace std;

// 手势映射到下标：0=B, 1=C, 2=J（正好与字母序 B<C<J 一致，便于并列取最小）
int idx(char c) {
    if (c == 'B') return 0;
    if (c == 'C') return 1;
    return 2;
}

int main() {
    int n;
    if (!(cin >> n)) return 0;

    int aWin = 0, aDraw = 0, aLose = 0;
    int aG[3] = {0}, bG[3] = {0}; // 各自以各手势获胜的次数

    for (int i = 0; i < n; ++i) {
        char a, b;
        cin >> a >> b;
        if (a == b) {
            ++aDraw;
        } else if ((a == 'B' && b == 'C') ||
                   (a == 'C' && b == 'J') ||
                   (a == 'J' && b == 'B')) {
            ++aWin;
            ++aG[idx(a)];
        } else {
            ++aLose;
            ++bG[idx(b)];
        }
    }

    const char order[3] = {'B', 'C', 'J'};
    auto best = [&](const int cnt[3]) {
        int k = 0;
        for (int i = 1; i < 3; ++i) if (cnt[i] > cnt[k]) k = i; // 严格大于才更新，保持字母序最小
        return order[k];
    };

    cout << aWin << ' ' << aDraw << ' ' << aLose << endl;
    cout << aLose << ' ' << aDraw << ' ' << aWin << endl;
    cout << best(aG) << ' ' << best(bG) << endl;
    return 0;
}

File: PAT_Basic_Level_by_java/1018 锤子剪刀布/1018 锤子剪刀布.java
/*
 * 题目编号：1018 锤子剪刀布
 * 题目描述：大家应该都会玩"锤子剪刀布"的游戏。现给出两人的交锋记录，请统计双方的胜、平、负次数，
 *   并给出双方分别用什么手势的胜算最大。
 *   规则：锤子（C）赢剪刀（J），剪刀（J）赢布（B），布（B）赢锤子（C）。
 *   如果获胜次数最多的手势不唯一，则输出字母序最小的那个（B < C < J）。
 * 实现原理：
 *   解题思路：
 *   - 读取交锋次数n，逐条记录判断胜负平。
 *   - 使用两个长度为3的数组aGest和bGest分别记录甲、乙获胜时使用的手势次数，
 *     下标映射：B→0, C→1, J→2。
 *   - 胜负判断：如果x == y为平局；否则检查甲是否获胜（三种胜局：C>J, J>B, B>C），
 *     如果甲胜则aWin++并记录手势，否则乙胜（即甲负）则aLose++并记录乙的手势。
 *   - 最后找出双方获胜次数最多的手势（同次数取字母序最小）。
 * 时间复杂度：O(n)，其中n为交锋次数。
 * 空间复杂度：O(1)，仅使用常量级额外空间。
 */
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

/**
 * 1018 锤子剪刀布
 * 统计甲乙胜负平，并给出各自获胜次数最多的手势（并列时取字母序最小 B<C<J）。
 */
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine().trim());
        int aWin = 0, aDraw = 0, aLose = 0;
        int[] aGest = new int[3]; // 甲获胜时使用的手势计数：0:B 1:C 2:J
        int[] bGest = new int[3]; // 乙获胜时使用的手势计数
        for (int i = 0; i < n; i++) {
            String[] p = br.readLine().trim().split("\\s+");
            char x = p[0].charAt(0), y = p[1].charAt(0);
            if (x == y) {
                aDraw++;                       // 平局
            } else if ((x == 'C' && y == 'J') || (x == 'J' && y == 'B') || (x == 'B' && y == 'C')) {
                aWin++;                        // 甲胜
                aGest[idx(x)]++;               // 记录甲获胜的手势
            } else {
                aLose++;                       // 乙胜（甲负）
                bGest[idx(y)]++;               // 记录乙获胜的手势
            }
        }
        // 输出甲的胜平负，以及乙的胜平负（乙的胜=甲的负，乙的负=甲的胜）
        System.out.println(aWin + " " + aDraw + " " + aLose);
        System.out.println(aLose + " " + aDraw + " " + aWin);
        // 输出双方获胜次数最多的手势
        System.out.println(best(aGest) + " " + best(bGest));
    }

    // 将手势字符映射到数组下标：B->0, C->1, J->2
    private static int idx(char c) {
        if (c == 'B') return 0;
        if (c == 'C') return 1;
        return 2;  // 'J'
    }

    // 找出数组中次数最多的手势，次数相同取字母序最小（B<C<J）
    private static char best(int[] g) {
        int max = -1, bi = 0;
        for (int i = 0; i < 3; i++) {
            if (g[i] > max) {
                max = g[i];
                bi = i;
            }
        }
        return bi == 0 ? 'B' : (bi == 1 ? 'C' : 'J');
    }
}